Source/DafnyCore/ProofDependencyWarnings.cs

```diff
@@ -42,15 +44,98 @@ public static void WarnAboutSuspiciousDependenciesForImplementation(DafnyOptions
     if (results.Any(r => r.Outcome != SolverOutcome.Valid)) {
       return;
     }
+    var unusedFunctions = UnusedFunctions(dafnyOptions, depManager, name, results.SelectMany(r => r.CoveredElements), results.Select(DafnyConsolePrinter.DistillVCResult).SelectMany(r => r.AvailableAxioms));
+    WarnAboutSuspiciousDependencies(dafnyOptions, reporter, depManager, name, results.Select(DafnyConsolePrinter.DistillVCResult).ToList(), unusedFunctions);
+  }
+
+  private static List<Function> UnusedFunctions(DafnyOptions dafnyOptions, ProofDependencyManager depManager, string name, IEnumerable<TrackedNodeComponent> coveredElements, IEnumerable<Axiom> axioms) {
+    if (!(dafnyOptions.Get(CommonOptionBag.SuggestProofRefactoring) && depManager.idsByMemberName[name].Decl is Method)) {
+      return new List<Function>();
+    }
+
+    var unusedFunctions = new List<Function>();
+    if (depManager.idsByMemberName[name].Decl is not Method m) {
+      return unusedFunctions;
+    }
+
+    var referencedFunctions = ReferencedFunctions(m);
+
+    var hiddenFunctions = HiddenFunctions(referencedFunctions);
+
+    var usedFunctions = coveredElements.Select(depManager.GetFullIdDependency).OfType<FunctionDefinitionDependency>()
+      .Select(dep => dep.function).Deduplicate((a, b) => a.Equals(b));
+
+    unusedFunctions = referencedFunctions.Except(hiddenFunctions).Except(usedFunctions).ToList();
```

Comment:
> unusedFunctions  = referencedFunctions.Except(hiddenFunctions).Except(usedFunctions)

Inside `HiddenFunctions`, based on `axioms` you implicitly compute `visibleFunctions` , and then define `hiddenFunctions = referencedFunctions.Except(visibleFunctions)`

so in total it adds up to 

```
unusedFunctions  = referencedFunctions.Except(referencedFunctions.Except(visibleFunctions)).Except(usedFunctions)
```

I'm guessing `a \ (a \ b)` simplifies to `a ∩ b`, so you could apply that to get:

```
unusedFunctions  = referencedFunctions.Intersect(visibleFunctions).Except(usedFunctions)
```

But then I'm still wondering whether you can't just do:
```
unusedFunctions  = visibleFunctions.Except(usedFunctions)
```

That'd be great because the computation for `referencedFunctions` looks rather expensive.

Reply:
Great suggestion, thanks! It is slightly different, in that it takes additional functions into account, but I like that change, too. Previously we only looked at the body, but not the spec.